src/axom/quest/IntersectionShaper.hpp

```diff
@@ -818,17 +1037,361 @@ class IntersectionShaper : public Shaper
+      GridFunctionView<ExecSpace> shapeVFView(shapeVolFrac);
+
+      // Workaround for HIP so we do not capture through "this" pointer.
+      const double* local_overlap_volumes = m_overlap_volumes;
+      const double* local_hex_volumes = m_hex_volumes;
+      axom::for_all<ExecSpace>(
+        dataSize,
+        AXOM_LAMBDA(axom::IndexType i) {
+          // Update this material's VF and vf_subtract, which is the
+          // amount to subtract from the gf's in updateVF.
+          double vf = (local_overlap_volumes[i] / local_hex_volumes[i]);
 
-    /// Implementation here -- update material volume fractions based on replacement rules
-    // Note: we're not yet updating the shape volume fractions
-    AXOM_UNUSED_VAR(shapeVolFrac);
+          // Write at most the writable amount.
+          double vf_actual = (vf <= vf_writable[i]) ? vf : vf_writable[i];
+
+          // NOTE: if matVFView[i] temporarily exceeds 1, it will be corrected
+          //       during the subtraction stage.
+          matVFView[i] += vf_actual;
+          vf_subtract[i] = vf_actual;
+
+          // Store the max shape VF.
+          shapeVFView[i] = vf;
+        });
+    });
+
+    // Iterate over updateVFs to subtract off VFs we allocated to the
+    // current shape's material.
+    AXOM_PERF_MARK_SECTION("update_vf", {
+      for(auto& gf : updateVFs)
+      {
+        GridFunctionView<ExecSpace> matVFView(gf);
+        axom::for_all<ExecSpace>(
+          dataSize,
+          AXOM_LAMBDA(axom::IndexType i) {
+            constexpr double INSIGNIFICANT_VOLFRAC = 1.e-14;
+            double s =
+              (matVFView[i] <= vf_subtract[i]) ? matVFView[i] : vf_subtract[i];
+            matVFView[i] -= s;
+            // Turn any slight negatives or positive insignificant volume fractions to zero.
+            matVFView[i] =
+              (matVFView[i] < INSIGNIFICANT_VOLFRAC) ? 0. : matVFView[i];
+            vf_subtract[i] -= s;
+          });
+      }
+    });
+
+    // Free temporary arrays
+    axom::deallocate(vf_subtract);
+    axom::deallocate(vf_writable);
+  }
+#endif
+
+  /*!
+   * \brief Apply material replacement rules for the current shape, using
+   *        the appropriate execution policy.
+   */
+  void applyReplacementRules(const klee::Shape& shape) override
+  {
+    switch(m_execPolicy)
+    {
+#if defined(AXOM_USE_RAJA) && defined(AXOM_USE_UMPIRE)
+    case seq:
+      applyReplacementRulesImpl<seq_exec>(shape);
+      break;
+  #if defined(AXOM_USE_OPENMP)
+    case omp:
+      applyReplacementRulesImpl<omp_exec>(shape);
+      break;
+  #endif  // AXOM_USE_OPENMP
+  #if defined(AXOM_USE_CUDA)
+    case cuda:
+      applyReplacementRulesImpl<cuda_exec>(shape);
+      break;
+  #endif  // AXOM_USE_CUDA
+  #if defined(AXOM_USE_HIP)
+    case hip:
+      applyReplacementRulesImpl<hip_exec>(shape);
+      break;
+  #endif  // AXOM_USE_HIP
+#endif    // AXOM_USE_RAJA && AXOM_USE_UMPIRE
```

Comment:
I think we've gotten warnings with some compilers about unused enum cases when the `#ifdef` guard covers the enum case.

A pattern like the following might work better:
```cpp
switch(m_execPolicy)
{
   ...
    case hip:
  #if defined(AXOM_USE_HIP)
      applyReplacementRulesImpl<hip_exec>(shape);
  #endif  // AXOM_USE_HIP
      break;
   ...
}
```

Reply:
I've seen such warnings when there is no "default" case present. I borrowed this pattern from existing methods such as prepareShapeQuery. These switches have a "default" case that nicely prints that the runtime policy is not handled. I'd prefer that to it silently doing nothing.